libs/datasource/src/datasource.cpp

```diff
@@ -0,0 +1,167 @@
+#include "httplib.h"
+#include "mapget/model/featurelayer.h"
+#include "mapget/model/stream.h"
+
+#include "datasource.h"
+
+#include <utility>
+#include <csignal>
+#include <atomic>
+
+namespace mapget
+{
+
+static std::atomic<DataSource*> activeDataSource;
+
+struct DataSource::Impl
+{
+    httplib::Server server_;
+    DataSourceInfo info_;
+    std::function<void(TileFeatureLayer&)> tileCallback_;
+    uint16_t port_ = 0;
+    std::thread serverThread_;
+    std::shared_ptr<Fields> fields_;
+
+    explicit Impl(DataSourceInfo info) : info_(std::move(info)), fields_(std::make_shared<Fields>(info_.nodeId_))
+    {
+        // Set up GET /tile endpoint
+        server_.Get(
+            "/tile",
+            [this](const httplib::Request& req, httplib::Response& res)
+            {
+                auto layerIdParam = req.get_param_value("layer");
+                auto layerIt = info_.layers_.find(layerIdParam);
+                if (layerIt == info_.layers_.end())
+                    throw std::runtime_error(stx::format("Unknown layer id `{}`!", layerIdParam));
+
+                auto tileIdParam = TileId{std::stoull(req.get_param_value("tileId"))};
+                auto fieldsOffsetParam = (simfil::FieldId)0;
+                if (req.has_param("fieldsOffset"))
+                    fieldsOffsetParam = (simfil::FieldId)
+                        std::stoul(req.get_param_value("fieldsOffset"));
+
+                auto tileFeatureLayer = std::make_shared<TileFeatureLayer>(
+                    tileIdParam,
+                    info_.nodeId_,
+                    info_.mapId_,
+                    layerIt->second,
+                    fields_);
+                if (tileCallback_)
+                    tileCallback_(*tileFeatureLayer);
+
+                // Serialize TileFeatureLayer using TileLayerStream
+                std::stringstream content;
+                TileLayerStream::FieldOffsetMap fieldOffsets{{info_.nodeId_, fieldsOffsetParam}};
+                TileLayerStream::Writer layerWriter{
+                    [&](auto&& msg) { content << msg; },
+                    fieldOffsets};
+                layerWriter.write(tileFeatureLayer);
+
+                res.set_content(content.str(), "application/binary");
+            });
+
+        // Set up GET /info endpoint
+        server_.Get(
+            "/info",
+            [this](const httplib::Request&, httplib::Response& res)
+            {
+                nlohmann::json j = info_.toJson();
+                res.set_content(j.dump(), "application/json");
+            });
+
+    }
+
+    static void handleSignal(int signal)
+    {
+        // Stop the active instance when a signal is received.
+        if (activeDataSource != nullptr) {
```

Comment:
Nitpick: Use `compare_exchange_strong` here to compare and reset in one atomic step.

Reply:
Interesting